pkg/apis/velero/v1/backup.go

```diff
@@ -24,76 +24,102 @@ import (
 	// should be included for consideration in the backup.
-	IncludeClusterResources *bool `json:"includeClusterResources"`
+	// +optional
+	IncludeClusterResources *bool `json:"includeClusterResources,omitempty"`
 
 	// Hooks represent custom behaviors that should be executed at different phases of the backup.
-	Hooks BackupHooks `json:"hooks"`
+	// +optional
+	Hooks BackupHooks `json:"hooks,omitempty"`
 
 	// StorageLocation is a string containing the name of a BackupStorageLocation where the backup should be stored.
-	StorageLocation string `json:"storageLocation"`
+	// +optional
+	StorageLocation string `json:"storageLocation,omitempty"`
 
 	// VolumeSnapshotLocations is a list containing names of VolumeSnapshotLocations associated with this backup.
-	VolumeSnapshotLocations []string `json:"volumeSnapshotLocations"`
+	// +optional
+	VolumeSnapshotLocations []string `json:"volumeSnapshotLocations,omitempty"`
 }
 
 // BackupHooks contains custom behaviors that should be executed at different phases of the backup.
 type BackupHooks struct {
 	// Resources are hooks that should be executed when backing up individual instances of a resource.
-	Resources []BackupResourceHookSpec `json:"resources"`
+	// +optional
+	Resources []BackupResourceHookSpec `json:"resources,omitempty"`
 }
 
 // BackupResourceHookSpec defines one or more BackupResourceHooks that should be executed based on
 // the rules defined for namespaces, resources, and label selector.
 type BackupResourceHookSpec struct {
 	// Name is the name of this hook.
 	Name string `json:"name"`
+
 	// IncludedNamespaces specifies the namespaces to which this hook spec applies. If empty, it applies
 	// to all namespaces.
-	IncludedNamespaces []string `json:"includedNamespaces"`
+	// +optional
+	IncludedNamespaces []string `json:"includedNamespaces,omitempty"`
+
 	// ExcludedNamespaces specifies the namespaces to which this hook spec does not apply.
-	ExcludedNamespaces []string `json:"excludedNamespaces"`
+	// +optional
+	ExcludedNamespaces []string `json:"excludedNamespaces,omitempty"`
+
 	// IncludedResources specifies the resources to which this hook spec applies. If empty, it applies
 	// to all resources.
-	IncludedResources []string `json:"includedResources"`
+	// +optional
+	IncludedResources []string `json:"includedResources,omitempty"`
+
 	// ExcludedResources specifies the resources to which this hook spec does not apply.
-	ExcludedResources []string `json:"excludedResources"`
+	// +optional
+	ExcludedResources []string `json:"excludedResources,omitempty"`
+
 	// LabelSelector, if specified, filters the resources to which this hook spec applies.
+	// +optional
 	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+
 	// PreHooks is a list of BackupResourceHooks to execute prior to storing the item in the backup.
 	// These are executed before any "additional items" from item actions are processed.
+	// +optional
```

Comment:
Is there a way to say "either PreHooks or PostHooks must be non-empty"?

Reply:
I think we'd need the "OneOf" validation in OpenAPIv3, but it doesn't look like Kubebuilder supports this (https://github.com/kubernetes-sigs/kubebuilder/issues/301, https://book.kubebuilder.io/reference/markers/crd-validation.html).

I'm actually not sure where Pre/PostHooks are used. Do you have any examples of them? Are you saying that if a PreHook is defined, a PostHook cannot be defined (and vice versa)?

Reply:
Got it - I'll read through those refs, thanks for the links.

I'll have to double-check the exact object structure here, but I think the semantics should be that there needs to be at least one pre or post hook - both is fine, but the slices can't both be empty simultaneously.

Reply:
We'll validate this in the controller anyway (I think), so if we can't do it with the structural schema it's OK - but if we can, it's good to add.

Reply:
It would be awesome if we could get the `OneOf` validation working so that the API server handles more of this validation for us...but if Kubebuilder doesn't support it yet, we may have to wait.

Reply:
> I'll have to double-check the exact object structure here, but I think the semantics should be that there needs to be at least one pre or post hook - both is fine, but the slices can't both be empty simultaneously.

@skriss hmm, I've definitely created backups before without any Pre or Post hooks? By default `velero create backup` sets `hooks: {}`.

Reply:
Yeah you can definitely create a backup without hooks. But, if you do create a hook, then it needs to have either a pre- or a post- hook component (or both).

Reply:
This isn't a must-have for getting this PR in, we can defer for now.